// @shopify/scripts-sdk-as/assembly/Money/Currency/Currencies.ts
// DO NOT MODIFY THIS FILE. This file was generated by "npm run build:currencies"

import {Currency} from './Currency';

export class Currencies {
  static AED: Currency = new Currency('AED', 100);
  static AFN: Currency = new Currency('AFN', 100);
  static ALL: Currency = new Currency('ALL', 100);
  static AMD: Currency = new Currency('AMD', 100);
  static ANG: Currency = new Currency('ANG', 100);
  static AOA: Currency = new Currency('AOA', 100);
  static ARS: Currency = new Currency('ARS', 100);
  static AUD: Currency = new Currency('AUD', 100);
  static AWG: Currency = new Currency('AWG', 100);
  static AZN: Currency = new Currency('AZN', 100);
  static BAM: Currency = new Currency('BAM', 100);
  static BBD: Currency = new Currency('BBD', 100);
  static BDT: Currency = new Currency('BDT', 100);
  static BGN: Currency = new Currency('BGN', 100);
  static BHD: Currency = new Currency('BHD', 1000);
  static BIF: Currency = new Currency('BIF', 1);
  static BMD: Currency = new Currency('BMD', 100);
  static BND: Currency = new Currency('BND', 100);
  static BOB: Currency = new Currency('BOB', 100);
  static BRL: Currency = new Currency('BRL', 100);
  static BSD: Currency = new Currency('BSD', 100);
  static BTN: Currency = new Currency('BTN', 100);
  static BWP: Currency = new Currency('BWP', 100);
  static BYN: Currency = new Currency('BYN', 100);
  static BYR: Currency = new Currency('BYR', 1);
  static BZD: Currency = new Currency('BZD', 100);
  static CAD: Currency = new Currency('CAD', 100);
  static CDF: Currency = new Currency('CDF', 100);
  static CHF: Currency = new Currency('CHF', 100);
  static CLF: Currency = new Currency('CLF', 10000);
  static CLP: Currency = new Currency('CLP', 1);
  static CNY: Currency = new Currency('CNY', 100);
  static COP: Currency = new Currency('COP', 100);
  static CRC: Currency = new Currency('CRC', 100);
  static CUC: Currency = new Currency('CUC', 100);
  static CUP: Currency = new Currency('CUP', 100);
  static CVE: Currency = new Currency('CVE', 100);
  static CZK: Currency = new Currency('CZK', 100);
  static DJF: Currency = new Currency('DJF', 1);
  static DKK: Currency = new Currency('DKK', 100);
  static DOP: Currency = new Currency('DOP', 100);
  static DZD: Currency = new Currency('DZD', 100);
  static EGP: Currency = new Currency('EGP', 100);
  static ERN: Currency = new Currency('ERN', 100);
  static ETB: Currency = new Currency('ETB', 100);
  static EUR: Currency = new Currency('EUR', 100);
  static FJD: Currency = new Currency('FJD', 100);
  static FKP: Currency = new Currency('FKP', 100);
  static GBP: Currency = new Currency('GBP', 100);
  static GEL: Currency = new Currency('GEL', 100);
  static GHS: Currency = new Currency('GHS', 100);
  static GIP: Currency = new Currency('GIP', 100);
  static GMD: Currency = new Currency('GMD', 100);
  static GNF: Currency = new Currency('GNF', 1);
  static GTQ: Currency = new Currency('GTQ', 100);
  static GYD: Currency = new Currency('GYD', 100);
  static HKD: Currency = new Currency('HKD', 100);
  static HNL: Currency = new Currency('HNL', 100);
  static HRK: Currency = new Currency('HRK', 100);
  static HTG: Currency = new Currency('HTG', 100);
  static HUF: Currency = new Currency('HUF', 100);
  static IDR: Currency = new Currency('IDR', 100);
  static ILS: Currency = new Currency('ILS', 100);
  static INR: Currency = new Currency('INR', 100);
  static IQD: Currency = new Currency('IQD', 1000);
  static IRR: Currency = new Currency('IRR', 100);
  static ISK: Currency = new Currency('ISK', 1);
  static JMD: Currency = new Currency('JMD', 100);
  static JOD: Currency = new Currency('JOD', 1000);
  static JPY: Currency = new Currency('JPY', 1);
  static KES: Currency = new Currency('KES', 100);
  static KGS: Currency = new Currency('KGS', 100);
  static KHR: Currency = new Currency('KHR', 100);
  static KMF: Currency = new Currency('KMF', 1);
  static KPW: Currency = new Currency('KPW', 100);
  static KRW: Currency = new Currency('KRW', 1);
  static KWD: Currency = new Currency('KWD', 1000);
  static KYD: Currency = new Currency('KYD', 100);
  static KZT: Currency = new Currency('KZT', 100);
  static LAK: Currency = new Currency('LAK', 100);
  static LBP: Currency = new Currency('LBP', 100);
  static LKR: Currency = new Currency('LKR', 100);
  static LRD: Currency = new Currency('LRD', 100);
  static LSL: Currency = new Currency('LSL', 100);
  static LTL: Currency = new Currency('LTL', 100);
  static LVL: Currency = new Currency('LVL', 100);
  static LYD: Currency = new Currency('LYD', 1000);
  static MAD: Currency = new Currency('MAD', 100);
  static MDL: Currency = new Currency('MDL', 100);
  static MGA: Currency = new Currency('MGA', 5);
  static MKD: Currency = new Currency('MKD', 100);
  static MMK: Currency = new Currency('MMK', 100);
  static MNT: Currency = new Currency('MNT', 100);
  static MOP: Currency = new Currency('MOP', 100);
  static MRO: Currency = new Currency('MRO', 5);
  static MUR: Currency = new Currency('MUR', 100);
  static MVR: Currency = new Currency('MVR', 100);
  static MWK: Currency = new Currency('MWK', 100);
  static MXN: Currency = new Currency('MXN', 100);
  static MYR: Currency = new Currency('MYR', 100);
  static MZN: Currency = new Currency('MZN', 100);
  static NAD: Currency = new Currency('NAD', 100);
  static NGN: Currency = new Currency('NGN', 100);
  static NIO: Currency = new Currency('NIO', 100);
  static NOK: Currency = new Currency('NOK', 100);
  static NPR: Currency = new Currency('NPR', 100);
  static NZD: Currency = new Currency('NZD', 100);
  static OMR: Currency = new Currency('OMR', 1000);
  static PAB: Currency = new Currency('PAB', 100);
  static PEN: Currency = new Currency('PEN', 100);
  static PGK: Currency = new Currency('PGK', 100);
  static PHP: Currency = new Currency('PHP', 100);
  static PKR: Currency = new Currency('PKR', 100);
  static PLN: Currency = new Currency('PLN', 100);
  static PYG: Currency = new Currency('PYG', 1);
  static QAR: Currency = new Currency('QAR', 100);
  static RON: Currency = new Currency('RON', 100);
  static RSD: Currency = new Currency('RSD', 100);
  static RUB: Currency = new Currency('RUB', 100);
  static RWF: Currency = new Currency('RWF', 1);
  static SAR: Currency = new Currency('SAR', 100);
  static SBD: Currency = new Currency('SBD', 100);
  static SCR: Currency = new Currency('SCR', 100);
  static SDG: Currency = new Currency('SDG', 100);
  static SEK: Currency = new Currency('SEK', 100);
  static SGD: Currency = new Currency('SGD', 100);
  static SHP: Currency = new Currency('SHP', 100);
  static SKK: Currency = new Currency('SKK', 100);
  static SLL: Currency = new Currency('SLL', 100);
  static SOS: Currency = new Currency('SOS', 100);
  static SRD: Currency = new Currency('SRD', 100);
  static SSP: Currency = new Currency('SSP', 100);
  static STD: Currency = new Currency('STD', 100);
  static SVC: Currency = new Currency('SVC', 100);
  static SYP: Currency = new Currency('SYP', 100);
  static SZL: Currency = new Currency('SZL', 100);
  static THB: Currency = new Currency('THB', 100);
  static TJS: Currency = new Currency('TJS', 100);
  static TMT: Currency = new Currency('TMT', 100);
  static TND: Currency = new Currency('TND', 1000);
  static TOP: Currency = new Currency('TOP', 100);
  static TRY: Currency = new Currency('TRY', 100);
  static TTD: Currency = new Currency('TTD', 100);
  static TWD: Currency = new Currency('TWD', 100);
  static TZS: Currency = new Currency('TZS', 100);
  static UAH: Currency = new Currency('UAH', 100);
  static UGX: Currency = new Currency('UGX', 1);
  static USD: Currency = new Currency('USD', 100);
  static UYU: Currency = new Currency('UYU', 100);
  static UZS: Currency = new Currency('UZS', 100);
  static VEF: Currency = new Currency('VEF', 100);
  static VND: Currency = new Currency('VND', 1);
  static VUV: Currency = new Currency('VUV', 1);
  static WST: Currency = new Currency('WST', 100);
  static XAF: Currency = new Currency('XAF', 1);
  static XAG: Currency = new Currency('XAG', 1);
  static XAU: Currency = new Currency('XAU', 1);
  static XBA: Currency = new Currency('XBA', 1);
  static XBB: Currency = new Currency('XBB', 1);
  static XBC: Currency = new Currency('XBC', 1);
  static XBD: Currency = new Currency('XBD', 1);
  static XCD: Currency = new Currency('XCD', 100);
  static XDR: Currency = new Currency('XDR', 1);
  static XOF: Currency = new Currency('XOF', 1);
  static XPD: Currency = new Currency('XPD', 1);
  static XPF: Currency = new Currency('XPF', 1);
  static XPT: Currency = new Currency('XPT', 1);
  static xts: Currency = new Currency('xts', 1);
  static YER: Currency = new Currency('YER', 100);
  static ZAR: Currency = new Currency('ZAR', 100);
  static ZMK: Currency = new Currency('ZMK', 100);
  static ZMW: Currency = new Currency('ZMW', 100);

  static fromCode(code: string): Currency {
    if (code == 'AED') return Currencies.AED;
    else if (code == 'AFN') return Currencies.AFN;
    else if (code == 'ALL') return Currencies.ALL;
    else if (code == 'AMD') return Currencies.AMD;
    else if (code == 'ANG') return Currencies.ANG;
    else if (code == 'AOA') return Currencies.AOA;
    else if (code == 'ARS') return Currencies.ARS;
    else if (code == 'AUD') return Currencies.AUD;
    else if (code == 'AWG') return Currencies.AWG;
    else if (code == 'AZN') return Currencies.AZN;
    else if (code == 'BAM') return Currencies.BAM;
    else if (code == 'BBD') return Currencies.BBD;
    else if (code == 'BDT') return Currencies.BDT;
    else if (code == 'BGN') return Currencies.BGN;
    else if (code == 'BHD') return Currencies.BHD;
    else if (code == 'BIF') return Currencies.BIF;
    else if (code == 'BMD') return Currencies.BMD;
    else if (code == 'BND') return Currencies.BND;
    else if (code == 'BOB') return Currencies.BOB;
    else if (code == 'BRL') return Currencies.BRL;
    else if (code == 'BSD') return Currencies.BSD;
    else if (code == 'BTN') return Currencies.BTN;
    else if (code == 'BWP') return Currencies.BWP;
    else if (code == 'BYN') return Currencies.BYN;
    else if (code == 'BYR') return Currencies.BYR;
    else if (code == 'BZD') return Currencies.BZD;
    else if (code == 'CAD') return Currencies.CAD;
    else if (code == 'CDF') return Currencies.CDF;
    else if (code == 'CHF') return Currencies.CHF;
    else if (code == 'CLF') return Currencies.CLF;
    else if (code == 'CLP') return Currencies.CLP;
    else if (code == 'CNY') return Currencies.CNY;
    else if (code == 'COP') return Currencies.COP;
    else if (code == 'CRC') return Currencies.CRC;
    else if (code == 'CUC') return Currencies.CUC;
    else if (code == 'CUP') return Currencies.CUP;
    else if (code == 'CVE') return Currencies.CVE;
    else if (code == 'CZK') return Currencies.CZK;
    else if (code == 'DJF') return Currencies.DJF;
    else if (code == 'DKK') return Currencies.DKK;
    else if (code == 'DOP') return Currencies.DOP;
    else if (code == 'DZD') return Currencies.DZD;
    else if (code == 'EGP') return Currencies.EGP;
    else if (code == 'ERN') return Currencies.ERN;
    else if (code == 'ETB') return Currencies.ETB;
    else if (code == 'EUR') return Currencies.EUR;
    else if (code == 'FJD') return Currencies.FJD;
    else if (code == 'FKP') return Currencies.FKP;
    else if (code == 'GBP') return Currencies.GBP;
    else if (code == 'GEL') return Currencies.GEL;
    else if (code == 'GHS') return Currencies.GHS;
    else if (code == 'GIP') return Currencies.GIP;
    else if (code == 'GMD') return Currencies.GMD;
    else if (code == 'GNF') return Currencies.GNF;
    else if (code == 'GTQ') return Currencies.GTQ;
    else if (code == 'GYD') return Currencies.GYD;
    else if (code == 'HKD') return Currencies.HKD;
    else if (code == 'HNL') return Currencies.HNL;
    else if (code == 'HRK') return Currencies.HRK;
    else if (code == 'HTG') return Currencies.HTG;
    else if (code == 'HUF') return Currencies.HUF;
    else if (code == 'IDR') return Currencies.IDR;
    else if (code == 'ILS') return Currencies.ILS;
    else if (code == 'INR') return Currencies.INR;
    else if (code == 'IQD') return Currencies.IQD;
    else if (code == 'IRR') return Currencies.IRR;
    else if (code == 'ISK') return Currencies.ISK;
    else if (code == 'JMD') return Currencies.JMD;
    else if (code == 'JOD') return Currencies.JOD;
    else if (code == 'JPY') return Currencies.JPY;
    else if (code == 'KES') return Currencies.KES;
    else if (code == 'KGS') return Currencies.KGS;
    else if (code == 'KHR') return Currencies.KHR;
    else if (code == 'KMF') return Currencies.KMF;
    else if (code == 'KPW') return Currencies.KPW;
    else if (code == 'KRW') return Currencies.KRW;
    else if (code == 'KWD') return Currencies.KWD;
    else if (code == 'KYD') return Currencies.KYD;
    else if (code == 'KZT') return Currencies.KZT;
    else if (code == 'LAK') return Currencies.LAK;
    else if (code == 'LBP') return Currencies.LBP;
    else if (code == 'LKR') return Currencies.LKR;
    else if (code == 'LRD') return Currencies.LRD;
    else if (code == 'LSL') return Currencies.LSL;
    else if (code == 'LTL') return Currencies.LTL;
    else if (code == 'LVL') return Currencies.LVL;
    else if (code == 'LYD') return Currencies.LYD;
    else if (code == 'MAD') return Currencies.MAD;
    else if (code == 'MDL') return Currencies.MDL;
    else if (code == 'MGA') return Currencies.MGA;
    else if (code == 'MKD') return Currencies.MKD;
    else if (code == 'MMK') return Currencies.MMK;
    else if (code == 'MNT') return Currencies.MNT;
    else if (code == 'MOP') return Currencies.MOP;
    else if (code == 'MRO') return Currencies.MRO;
    else if (code == 'MUR') return Currencies.MUR;
    else if (code == 'MVR') return Currencies.MVR;
    else if (code == 'MWK') return Currencies.MWK;
    else if (code == 'MXN') return Currencies.MXN;
    else if (code == 'MYR') return Currencies.MYR;
    else if (code == 'MZN') return Currencies.MZN;
    else if (code == 'NAD') return Currencies.NAD;
    else if (code == 'NGN') return Currencies.NGN;
    else if (code == 'NIO') return Currencies.NIO;
    else if (code == 'NOK') return Currencies.NOK;
    else if (code == 'NPR') return Currencies.NPR;
    else if (code == 'NZD') return Currencies.NZD;
    else if (code == 'OMR') return Currencies.OMR;
    else if (code == 'PAB') return Currencies.PAB;
    else if (code == 'PEN') return Currencies.PEN;
    else if (code == 'PGK') return Currencies.PGK;
    else if (code == 'PHP') return Currencies.PHP;
    else if (code == 'PKR') return Currencies.PKR;
    else if (code == 'PLN') return Currencies.PLN;
    else if (code == 'PYG') return Currencies.PYG;
    else if (code == 'QAR') return Currencies.QAR;
    else if (code == 'RON') return Currencies.RON;
    else if (code == 'RSD') return Currencies.RSD;
    else if (code == 'RUB') return Currencies.RUB;
    else if (code == 'RWF') return Currencies.RWF;
    else if (code == 'SAR') return Currencies.SAR;
    else if (code == 'SBD') return Currencies.SBD;
    else if (code == 'SCR') return Currencies.SCR;
    else if (code == 'SDG') return Currencies.SDG;
    else if (code == 'SEK') return Currencies.SEK;
    else if (code == 'SGD') return Currencies.SGD;
    else if (code == 'SHP') return Currencies.SHP;
    else if (code == 'SKK') return Currencies.SKK;
    else if (code == 'SLL') return Currencies.SLL;
    else if (code == 'SOS') return Currencies.SOS;
    else if (code == 'SRD') return Currencies.SRD;
    else if (code == 'SSP') return Currencies.SSP;
    else if (code == 'STD') return Currencies.STD;
    else if (code == 'SVC') return Currencies.SVC;
    else if (code == 'SYP') return Currencies.SYP;
    else if (code == 'SZL') return Currencies.SZL;
    else if (code == 'THB') return Currencies.THB;
    else if (code == 'TJS') return Currencies.TJS;
    else if (code == 'TMT') return Currencies.TMT;
    else if (code == 'TND') return Currencies.TND;
    else if (code == 'TOP') return Currencies.TOP;
    else if (code == 'TRY') return Currencies.TRY;
    else if (code == 'TTD') return Currencies.TTD;
    else if (code == 'TWD') return Currencies.TWD;
    else if (code == 'TZS') return Currencies.TZS;
    else if (code == 'UAH') return Currencies.UAH;
    else if (code == 'UGX') return Currencies.UGX;
    else if (code == 'USD') return Currencies.USD;
    else if (code == 'UYU') return Currencies.UYU;
    else if (code == 'UZS') return Currencies.UZS;
    else if (code == 'VEF') return Currencies.VEF;
    else if (code == 'VND') return Currencies.VND;
    else if (code == 'VUV') return Currencies.VUV;
    else if (code == 'WST') return Currencies.WST;
    else if (code == 'XAF') return Currencies.XAF;
    else if (code == 'XAG') return Currencies.XAG;
    else if (code == 'XAU') return Currencies.XAU;
    else if (code == 'XBA') return Currencies.XBA;
    else if (code == 'XBB') return Currencies.XBB;
    else if (code == 'XBC') return Currencies.XBC;
    else if (code == 'XBD') return Currencies.XBD;
    else if (code == 'XCD') return Currencies.XCD;
    else if (code == 'XDR') return Currencies.XDR;
    else if (code == 'XOF') return Currencies.XOF;
    else if (code == 'XPD') return Currencies.XPD;
    else if (code == 'XPF') return Currencies.XPF;
    else if (code == 'XPT') return Currencies.XPT;
    else if (code == 'xts') return Currencies.xts;
    else if (code == 'YER') return Currencies.YER;
    else if (code == 'ZAR') return Currencies.ZAR;
    else if (code == 'ZMK') return Currencies.ZMK;
    else if (code == 'ZMW') return Currencies.ZMW;

    throw new Error(code + ' is not a known currency.');
  }
}
